app/core/core-app-review/src/main/kotlin/com/hedvig/android/core/appreview/WaitUntilAppReviewDialogShouldBeOpenedUseCase.kt

```diff
@@ -0,0 +1,25 @@
+package com.hedvig.android.core.appreview
+
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+
+/**
+ * This function will suspend and wait forever until the first time we should open the app review dialog.
+ */
+interface WaitUntilAppReviewDialogShouldBeOpenedUseCase {
+  suspend fun invoke(): Unit
+}
+
+internal class WaitUntilAppReviewDialogShouldBeOpenedUseCaseImpl(
+  private val dataStore: SelfServiceCompletedEventStore,
+) : WaitUntilAppReviewDialogShouldBeOpenedUseCase {
+  override suspend fun invoke() {
+    dataStore.observeNumberOfCompletedSelfServices()
+      // Drop the first number, whatever it was, to not show a dialog right as the app is launched. We are only
+      // interested in observing changes to this number after we've already started observing it
+      .drop(1)
```

Comment:
That drop would only be used if we by some reason haven't showed the review dialog last time the app was opened  and the member completed selfService, right?

Reply:
Hmm not quite, it's a bit of a complicated situation but I can try to explain.
The place in which we are listening in on this flow is wrapped in a `repeatOnLifecycle(Lifecycle.State.STARTED)`, since we only want to ever call this if the app is in the foreground.
So that means that if someone goes to home, and comes back into the app, this will in fact be subscribed to again from scratch.
The DataStore, when a new observer is trying to observe a value from inside of it, will on the flow it gives you emit the first value that it has already in the disk, and then it will wait until any changes happen so that it emits again.
So in short, this `drop(1)` will simply ignore the first emission from the DataStore since that one is just what was on the disk itself, we do not care for what it already was. We only want to know that if it has *changed*, which may only happen if one was in the app already, and the `selfServiceCompletedEventManager.completedSelfServiceSuccessfully()` was called, so the counter has gone up, so the datastore will emit another value, which will continue further down in this flow chain, since `drop(1)` will have already happened.

Reply:
Yup the in-app review APIs handle this automatically for us. https://developer.android.com/guide/playcore/in-app-review#quotas
"To provide a great user experience, Google Play enforces a time-bound quota on how often a user can be shown the review dialog. Because of this quota, calling the launchReviewFlow method more than once during a short period of time (for example, less than a month) might not always display a dialog."